pkg/buganalysis/cache.go

```diff
@@ -71,15 +71,18 @@ func NewBugCache() BugCache {
 
 // updates a global variable with the bug mapping based on current failures.
 func (c *bugCache) UpdateForFailedTests(failedTestNames ...string) error {
-	newBugs, lastUpdateError := findBugsForFailedTests(failedTestNames...)
+	newFailedTestNames := []string{}
+	for _, testName := range failedTestNames {
+		if _, found := c.cache[testName]; !found {
```

Comment:
to be "safe" this needs to be inside the lock.

Reply:
cool, wasn't sure if read operations needed to be inside the lock. the `findBugsForFailedTests` function can be quite slow, should I lock/unlock for reads, run that function, then lock/defer unlock afterward? Presumably we don't want to lock for 10s of seconds while we wait for bug lookups

Reply:
take a readlock on it, like ListBugs does.

readLock
  check cache, accumulate list of names not in the cache
unreadlock

bugquery

writelock
  update cache w/ query results
unwritelock

i think that achieves the goal.